### PR TITLE
should have a config option for stun server

### PR DESCRIPTION
**Pull request by RyanKung**

### CODE REVIEW

**Comment by RyanKung** [comment]
About Http services

/// 1. GET /sdp
/// Which create offer and send back to candidated peer
/// 2. POST /sdp
/// Which receive offer from peer and send the answer back
/// 2. Get /connect/{url}
/// Which request offer from peer and sand the answer back

/// SDP Scheme:
/// Browser -> Get offer from server
/// Server -> Send Offer and set it as local_description (implemented in transport)
/// Browser -> Send answer back to server, Server set it to remote description

/// SDP Forward Scheme:
/// Server A -> Requset offer from Server B, and set it as remote_descriton
/// Server A -> sent local_desc as answer to Server B

**Comment by RyanKung** [comment]
```
WARN - [controlling]: could not listen udp fe:80::: io error: Can't assign requested address (os error 49)
```

TODO: Disable IPV6 by default

**Comment by RyanKung** [comment]
Test with `curl -v  http://127.0.0.1:8964/connect\?node\=http://127.0.0.1:8965/sdp`


Peer A:
```
➜  bns-node git:(features/bugfix_for_stun_etc) ✗ cargo run -- -d 127.0.0.1:8964
    Finished dev [unoptimized + debuginfo] target(s) in 0.13s
     Running `target/debug/bns-node -d '127.0.0.1:8964'`
Serving on 127.0.0.1:8964
INFO - receive request to GET /connect
INFO - v=0
o=- 1909452493917867017 619681000 IN IP4 0.0.0.0
s=-
t=0 0
a=fingerprint:sha-256 F8:AA:30:44:CD:28:D0:B2:F9:67:8F:76:9A:26:56:A7:64:09:FA:DB:A4:9D:CE:BC:ED:54:6D:AE:AA:1B:F8:80
a=group:BUNDLE 0
m=application 9 UDP/DTLS/SCTP webrtc-datachannel
c=IN IP4 0.0.0.0
a=setup:actpass
a=mid:0
a=sendrecv
a=sctp-port:5000
a=ice-ufrag:ULSxZjUEIvIsGhcF
a=ice-pwd:YYBvUteajmyLWTXKXxotofcDDELPJiuQ

INFO - Pending transport not exists, creating new
INFO - mDNS is using 0.0.0.0:5353 as dest_addr
INFO - Looping and listening Ok(0.0.0.0:5353)
INFO - try set_local_descrition as: RTCSessionDescription { sdp_type: Offer, sdp: "v=0\r\no=- 6764606610103535416 695529000 IN IP4 0.0.0.0\r\ns=-\r\nt=0 0\r\na=fingerprint:sha-256 36:E5:BB:33:D1:34:DF:75:14:92:A0:41:3A:F6:95:86:9E:17:9A:82:DA:BF:1E:15:D3:39:CB:6E:2D:61:B8:E6\r\na=group:BUNDLE 0\r\nm=application 9 UDP/DTLS/SCTP webrtc-datachannel\r\nc=IN IP4 0.0.0.0\r\na=setup:actpass\r\na=mid:0\r\na=sendrecv\r\na=sctp-port:5000\r\na=ice-ufrag:dXrCPGHLfkNxEOnh\r\na=ice-pwd:VIfTuZDcRQLyiHmcfBVUPqrGIgfeoVZM\r\n", parsed: Some(SessionDescription { version: 0, origin: Origin { username: "-", session_id: 6764606610103535416, session_version: 695529000, network_type: "IN", address_type: "IP4", unicast_address: "0.0.0.0" }, session_name: "-", session_information: None, uri: None, email_address: None, phone_number: None, connection_information: None, bandwidth: [], time_descriptions: [TimeDescription { timing: Timing { start_time: 0, stop_time: 0 }, repeat_times: [] }], time_zones: [], encryption_key: None, attributes: [Attribute { key: "fingerprint", value: Some("sha-256 36:E5:BB:33:D1:34:DF:75:14:92:A0:41:3A:F6:95:86:9E:17:9A:82:DA:BF:1E:15:D3:39:CB:6E:2D:61:B8:E6") }, Attribute { key: "group", value: Some("BUNDLE 0") }], media_descriptions: [MediaDescription { media_name: MediaName { media: "application", port: RangedPort { value: 9, range: None }, protos: ["UDP", "DTLS", "SCTP"], formats: ["webrtc-datachannel"] }, media_title: None, connection_information: Some(ConnectionInformation { network_type: "IN", address_type: "IP4", address: Some(Address { address: "0.0.0.0", ttl: None, range: None }) }), bandwidth: [], encryption_key: None, attributes: [Attribute { key: "setup", value: Some("actpass") }, Attribute { key: "mid", value: Some("0") }, Attribute { key: "sendrecv", value: None }, Attribute { key: "sctp-port:5000", value: None }, Attribute { key: "ice-ufrag", value: Some("dXrCPGHLfkNxEOnh") }, Attribute { key: "ice-pwd", value: Some("VIfTuZDcRQLyiHmcfBVUPqrGIgfeoVZM") }] }] }) }
INFO - signaling state changed to have-local-offer
WARN - [controlled]: could not listen udp fe:80::: io error: Can't assign requested address (os error 49)
WARN - [controlled]: failed to resolve stun host: stun.l.google.com:19302: io error: No available ipv6 IP address found!
INFO - register data channel callback
INFO - setting on message callback
INFO - assigning new transport
INFO - try set_remote_descrition as: RTCSessionDescription { sdp_type: Answer, sdp: "v=0\r\no=- 1909452493917867017 619681000 IN IP4 0.0.0.0\r\ns=-\r\nt=0 0\r\na=fingerprint:sha-256 F8:AA:30:44:CD:28:D0:B2:F9:67:8F:76:9A:26:56:A7:64:09:FA:DB:A4:9D:CE:BC:ED:54:6D:AE:AA:1B:F8:80\r\na=group:BUNDLE 0\r\nm=application 9 UDP/DTLS/SCTP webrtc-datachannel\r\nc=IN IP4 0.0.0.0\r\na=setup:actpass\r\na=mid:0\r\na=sendrecv\r\na=sctp-port:5000\r\na=ice-ufrag:ULSxZjUEIvIsGhcF\r\na=ice-pwd:YYBvUteajmyLWTXKXxotofcDDELPJiuQ\r\n", parsed: None }
INFO - signaling state changed to stable
INFO - [controlling]: Setting new connection state: Checking
INFO - ICE connection state changed: checking
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
```

Peer B: 
```
➜  bns-node git:(features/bugfix_for_stun_etc) ✗ cargo run -- -d 127.0.0.1:8965
    Finished dev [unoptimized + debuginfo] target(s) in 0.12s
     Running `target/debug/bns-node -d '127.0.0.1:8965'`
Serving on 127.0.0.1:8965
INFO - receive request to GET /sdp
INFO - Pending transport not exists, creating new
INFO - mDNS is using 0.0.0.0:5353 as dest_addr
INFO - Looping and listening Ok(0.0.0.0:5353)
INFO - try set_local_descrition as: RTCSessionDescription { sdp_type: Offer, sdp: "v=0\r\no=- 1909452493917867017 619681000 IN IP4 0.0.0.0\r\ns=-\r\nt=0 0\r\na=fingerprint:sha-256 F8:AA:30:44:CD:28:D0:B2:F9:67:8F:76:9A:26:56:A7:64:09:FA:DB:A4:9D:CE:BC:ED:54:6D:AE:AA:1B:F8:80\r\na=group:BUNDLE 0\r\nm=application 9 UDP/DTLS/SCTP webrtc-datachannel\r\nc=IN IP4 0.0.0.0\r\na=setup:actpass\r\na=mid:0\r\na=sendrecv\r\na=sctp-port:5000\r\na=ice-ufrag:ULSxZjUEIvIsGhcF\r\na=ice-pwd:YYBvUteajmyLWTXKXxotofcDDELPJiuQ\r\n", parsed: Some(SessionDescription { version: 0, origin: Origin { username: "-", session_id: 1909452493917867017, session_version: 619681000, network_type: "IN", address_type: "IP4", unicast_address: "0.0.0.0" }, session_name: "-", session_information: None, uri: None, email_address: None, phone_number: None, connection_information: None, bandwidth: [], time_descriptions: [TimeDescription { timing: Timing { start_time: 0, stop_time: 0 }, repeat_times: [] }], time_zones: [], encryption_key: None, attributes: [Attribute { key: "fingerprint", value: Some("sha-256 F8:AA:30:44:CD:28:D0:B2:F9:67:8F:76:9A:26:56:A7:64:09:FA:DB:A4:9D:CE:BC:ED:54:6D:AE:AA:1B:F8:80") }, Attribute { key: "group", value: Some("BUNDLE 0") }], media_descriptions: [MediaDescription { media_name: MediaName { media: "application", port: RangedPort { value: 9, range: None }, protos: ["UDP", "DTLS", "SCTP"], formats: ["webrtc-datachannel"] }, media_title: None, connection_information: Some(ConnectionInformation { network_type: "IN", address_type: "IP4", address: Some(Address { address: "0.0.0.0", ttl: None, range: None }) }), bandwidth: [], encryption_key: None, attributes: [Attribute { key: "setup", value: Some("actpass") }, Attribute { key: "mid", value: Some("0") }, Attribute { key: "sendrecv", value: None }, Attribute { key: "sctp-port:5000", value: None }, Attribute { key: "ice-ufrag", value: Some("ULSxZjUEIvIsGhcF") }, Attribute { key: "ice-pwd", value: Some("YYBvUteajmyLWTXKXxotofcDDELPJiuQ") }] }] }) }
INFO - signaling state changed to have-local-offer
WARN - [controlled]: could not listen udp fe:80::: io error: Can't assign requested address (os error 49)
WARN - [controlled]: failed to resolve stun host: stun.l.google.com:19302: io error: No available ipv6 IP address found!
INFO - register data channel callback
INFO - setting on message callback
INFO - assigning new transport
INFO - receive request to POST /sdp
INFO - Pending transport exists
INFO - try set_remote_descrition as: RTCSessionDescription { sdp_type: Answer, sdp: "v=0\r\no=- 6764606610103535416 695529000 IN IP4 0.0.0.0\r\ns=-\r\nt=0 0\r\na=fingerprint:sha-256 36:E5:BB:33:D1:34:DF:75:14:92:A0:41:3A:F6:95:86:9E:17:9A:82:DA:BF:1E:15:D3:39:CB:6E:2D:61:B8:E6\r\na=group:BUNDLE 0\r\nm=application 9 UDP/DTLS/SCTP webrtc-datachannel\r\nc=IN IP4 0.0.0.0\r\na=setup:actpass\r\na=mid:0\r\na=sendrecv\r\na=sctp-port:5000\r\na=ice-ufrag:dXrCPGHLfkNxEOnh\r\na=ice-pwd:VIfTuZDcRQLyiHmcfBVUPqrGIgfeoVZM\r\n", parsed: None }
INFO - signaling state changed to stable
INFO - set pending transport to None
INFO - [controlling]: Setting new connection state: Checking
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
INFO - ICE connection state changed: checking
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
````

**Comment by RyanKung** [comment]
Still

INFO - ICE connection state changed: checking
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.

**Comment by RyanKung** [comment]
The warning log is from here:
https://github.com/webrtc-rs/ice/blob/92cc698a3dc6da459f3bf3789fd046c2dffdf107/src/agent/agent_internal.rs

**Comment by RyanKung** [comment]
TRACE - start_transports: ice_role=controlling, dtls_role=auto
DEBUG - Started agent: isControlling? true, remoteUfrag: AXUZVkYWHTVAaAxn, remotePwd: oFCIiUqzNJSBYEvYeWnBhXkNabsxDaey
INFO - [controlling]: Setting new connection state: Checking
INFO - ICE connection state changed: checking
TRACE - [controlling]: pinging all candidates
WARN - [controlling]: pingAllCandidates called with no candidate pairs. Connection is not possible yet.
TRACE - signal: Want
TRACE - signal found waiting giver, notifying
TRACE - poll_want: taker wants!
TRACE - signal: Want
TRACE - signal: Want

**Comment by RyanKung** [comment]
For debug

```
cargo run -- -d 127.0.0.1:50000 -v Trace
cargo run -- -d 127.0.0.1:50001 -v Trace
url -v  http://127.0.0.1:8964/connect\?node\=http://127.0.0.1:50000/sdp
```

**Comment by Raytlty** [comment]
curl -v -H 'Content-Type: application/json' -d '{"node": "http://127.0.0.1:50001/offer"}' -X POST http://127.0.0.1:50000/connect

Update

File: bns-core/src/swarm/mod.rs [hunk 1]
```diff
@@ -47,8 +47,22 @@ impl Swarm {
     pub async fn new_transport(&mut self) -> Result<Arc<Transport>> {
         let mut ice_transport = Transport::new(Arc::clone(&self.signaler));
         ice_transport.start(self.stun_server.to_owned()).await?;
-        ice_transport.setup_callback().await?;
```
Comment (RyanKung) [inline bns-core/src/swarm/mod.rs:48]: why unbox?

File: src/discoveries/http.rs [hunk 2]
```diff
@@ -18,92 +15,90 @@ use bns_core::swarm::Swarm;
...
+                .send()
+                .await
+                .unwrap();
+            let sdp_str = resp.text().await.unwrap();
+            let answer = serde_json::from_str::<RTCSessionDescription>(&sdp_str).unwrap();
+            transport.set_remote_description(answer).await.unwrap();
+            transport.add_remote_server(&host_and_ip).await.unwrap();
             Response::builder().status(200).body(Body::empty())
         }
+        (&Method::POST, "/candidate") => {
```
Comment (RyanKung) [inline src/discoveries/http.rs]: `send candidate` via http? I thought it can be done var p2p

**Pull request closed without merging**